pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotRealtimeSegmentManager.java

```diff
@@ -138,15 +138,22 @@ private synchronized void assignRealtimeSegmentsToServerInstancesIfNecessary()
         for (String instanceId : instancesInResource) {
           InstanceZKMetadata instanceZKMetadata = _pinotHelixResourceManager.getInstanceZKMetadata(instanceId);
           String groupId = instanceZKMetadata.getGroupId(resource);
+          String partitionId = instanceZKMetadata.getPartition(resource);
+          AbstractTableConfig realtimeTableConfig =
+              ZKMetadataProvider.getRealtimeTableConfig(_pinotHelixResourceManager.getPropertyStore(), resource);
+          if (instancesInResource.size() == Integer.parseInt(realtimeTableConfig.getValidationConfig().getReplication())) {
```

Comment:
I don't like this special case. Say that for some reason, we wanted to change the replication level of a table, it's two operations (update property store, update ideal state). However, since we override the partition id based on ideal state and replication factor being identical, this would mean that behavior would change between the two operations.

We'd be better off requiring all instances to have a valid partition id in their instance metadata.

I haven't checked if we default to assigning "ALL" as the partition id in instance metadata, but if we don't, we should assign them the "ALL" partition id when we compute their initial consumer group id.